File: app/src/main/java/com/bookyrself/bookyrself/presenters/ContactsFragmentPresenter.java
package com.bookyrself.bookyrself.presenters;

import com.bookyrself.bookyrself.data.Contacts.ContactsRepository;
import com.bookyrself.bookyrself.data.ResponseModels.User.User;
import com.bookyrself.bookyrself.views.MainActivity;
import com.google.firebase.auth.FirebaseAuth;

import java.util.NoSuchElementException;

import io.reactivex.disposables.CompositeDisposable;

public class ContactsFragmentPresenter implements BasePresenter {
    private final ContactsPresenterListener presenterListener;
    private final ContactsRepository contactsRepository;
    private final CompositeDisposable compositeDisposable;
    private String userId;

    /**
     * Constructor
     */
    public ContactsFragmentPresenter(ContactsPresenterListener listener) {
        this.presenterListener = listener;
        this.contactsRepository = MainActivity.getContactsRepo();
        this.compositeDisposable = new CompositeDisposable();
    }

    /**
     * Methods
     */
    private void loadContacts() {
            compositeDisposable
                    .add(contactsRepository.getContactsForUser(userId)
                            .subscribe(
                                    stringUserEntry ->
                                            presenterListener.contactReturned(stringUserEntry.getKey(), stringUserEntry.getValue()),
                                    throwable -> {
                                        if (throwable instanceof NoSuchElementException) {
                                            presenterListener.noContactsReturned();
                                        } else {
                                            presenterListener.presentError(throwable.getMessage());
                                        }
                                    }));
        }


    @Override
    public void subscribe() {
        if (FirebaseAuth.getInstance().getUid() != null) {
            userId = FirebaseAuth.getInstance().getUid();
            loadContacts();
        } else {
            presenterListener.showSignedOutEmptyState();
        }

    }

    @Override
    public void unsubscribe() {
        compositeDisposable.clear();
    }


    /**
     * PresenterListener Definition
     */
    public interface ContactsPresenterListener extends BasePresenterListener {

        void noContactsReturned();

        void contactReturned(String id, User user);

        void presentError(String error);

    }
}

File: app/src/main/java/com/bookyrself/bookyrself/presenters/ContactsFragmentPresenter.kt
package com.bookyrself.bookyrself.presenters

import com.bookyrself.bookyrself.data.Contacts.ContactsRepository
import com.bookyrself.bookyrself.data.ServerModels.User.User
import com.bookyrself.bookyrself.views.MainActivity
import com.google.firebase.auth.FirebaseAuth
import io.reactivex.disposables.CompositeDisposable
import java.util.*

class ContactsFragmentPresenter(private val presenterListener: ContactsPresenterListener) : BasePresenter {

    private val contactsRepository: ContactsRepository = MainActivity.getContactsRepo()
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    private var userId: String? = null

    /**
     * Methods
     */
    private fun loadContacts() {
        compositeDisposable
                .add(contactsRepository.getContactsForUser(userId)
                        .subscribe(
                                { stringUserEntry -> presenterListener.contactReturned(stringUserEntry.key, stringUserEntry.value) },
                                { throwable ->
                                    if (throwable is NoSuchElementException) {
                                        presenterListener.noContactsReturned()
                                    } else {
                                        throwable.message?.let { presenterListener.presentError(it) }
                                    }
                                }))
    }


    override fun subscribe() {
        if (FirebaseAuth.getInstance().uid != null) {
            userId = FirebaseAuth.getInstance().uid
            loadContacts()
        } else {
            presenterListener.showSignedOutEmptyState()
        }

    }

    override fun unsubscribe() {
        compositeDisposable.clear()
    }


    /**
     * PresenterListener Definition
     */
    interface ContactsPresenterListener : BasePresenterListener {

        fun noContactsReturned()

        fun contactReturned(id: String, user: User)

        fun presentError(error: String)

    }
}
